node/src/tasks/bsp_charge_fees.rs

```diff
@@ -104,3 +198,197 @@ where
+    }
+}
+
+impl<FL, FSH> EventHandler<SpStopStoringInsolventUser> for BspChargeFeesTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: BspForestStorageHandlerT,
+{
+    async fn handle_event(&mut self, event: SpStopStoringInsolventUser) -> anyhow::Result<()> {
+        info!(
+            target: LOG_TARGET,
+            "Processing SpStopStoringForInsolventUser for user {:?}",
+            event.owner
+        );
+
+        // Get the insolvent user from the event.
+        let insolvent_user = event.owner;
+
+        // Queue a request to confirm the storing of the file.
+        self.storage_hub_handler
+            .blockchain
+            .queue_stop_storing_for_insolvent_user_request(StopStoringForInsolventUserRequest::new(
+                insolvent_user,
+            ))
+            .await?;
+
+        Ok(())
+    }
+}
+
+/// Handles the `ProcessStopStoringForInsolventUserRequest` event.
+///
+/// This event is triggered whenever a Forest write-lock can be acquired to process a `StopStoringForInsolventUserRequest`
+/// after receiving either a `UserWithoutFunds` or `SpStopStoringInsolventUser` event.
+impl<FL, FSH> EventHandler<ProcessStopStoringForInsolventUserRequest> for BspChargeFeesTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: BspForestStorageHandlerT,
+{
+    async fn handle_event(
+        &mut self,
+        event: ProcessStopStoringForInsolventUserRequest,
+    ) -> anyhow::Result<()> {
+        info!(
+            target: LOG_TARGET,
+            "Processing StopStoringForInsolventUserRequest for user: {:?}",
+            event.data.who,
+        );
+
+        // Get the insolvent user from the event.
+        let insolvent_user = event.data.who;
+
+        // Get a write-lock on the forest root since we are going to be modifying it by removing a user's file.
+        let forest_root_write_tx = match event.forest_root_write_tx.lock().await.take() {
+            Some(tx) => tx,
+            None => {
+                error!(target: LOG_TARGET, "CRITICAL❗️❗️ This is a bug! Forest root write tx already taken. This is a critical bug. Please report it to the StorageHub team.");
+                return Err(anyhow!(
+                    "CRITICAL❗️❗️ This is a bug! Forest root write tx already taken!"
+                ));
+            }
+        };
+
+        // Get the forest storage.
+        let fs = self
+            .storage_hub_handler
+            .forest_storage_handler
+            .get(&NoKey)
+            .await
+            .ok_or_else(|| anyhow!("Failed to get forest storage."))?;
+
+        // Get all the files that belong to the insolvent user, delete the first one.
+        let user_files = fs
+            .read()
+            .await
+            .get_files_by_user(&insolvent_user)
+            .map_err(|e| anyhow!("Failed to get metadata from Forest: {:?}", e))?;
+
+        if !user_files.is_empty() {
+            let (file_key, metadata) = user_files.first().expect("User files is not empty");
```

Comment:
Are you just deleting the first file?

Reply:
Yep, this task deletes one file from the user each block until it has correctly deleted all its files, to avoid having multiple root-changing transactions in one block

Reply:
I see. So it would be ideal that we make a batch deletion transaction, right?

Reply:
100%, the same way we have a batch for BSPs confirming to store a file we should have batching for deletion as well.